ext/RMagick/extconf.rb

```diff
@@ -123,31 +123,73 @@ def exit_failure(msg)
       exit(1)
     end
 
+    def detect_imagemagick_packages(packages)
+      existed_packages = packages.map do |package|
+        if system "pkg-config --exists #{package}"
+          package
+        end
+      end
+      existed_packages.compact
+    end
+
+    def detect_im6_packages
+      packages = %w[
+        ImageMagick-6.Q64HDRI
+        ImageMagick-6.Q32HDRI
+        ImageMagick-6.Q16HDRI
+        ImageMagick-6.Q8HDRI
+        ImageMagick-6.Q64
+        ImageMagick-6.Q32
+        ImageMagick-6.Q16
+        ImageMagick-6.Q8
+      ]
+      @detect_im6_packages ||= detect_imagemagick_packages(packages)
+    end
+
+    def detect_im7_packages
+      packages = %w[
+        ImageMagick-7.Q64HDRI
+        ImageMagick-7.Q32HDRI
+        ImageMagick-7.Q16HDRI
+        ImageMagick-7.Q8HDRI
+        ImageMagick-7.Q64
+        ImageMagick-7.Q32
+        ImageMagick-7.Q16
+        ImageMagick-7.Q8
+      ]
+      @detect_im7_packages ||= detect_imagemagick_packages(packages)
+    end
+
+    def im6_package_exists?
+      package = detect_im6_packages
+      package.empty? ? false : true
+    end
+
+    def im7_package_exists?
+      package = detect_im7_packages
+      package.empty? ? false : true
+    end
+
     def determine_imagemagick_package
-      packages = `pkg-config --list-all`.scan(/(ImageMagick\-[\.A-Z0-9]+) .*/).flatten
+      packages = [detect_im7_packages, detect_im6_packages].flatten
 
-      # For ancient version of ImageMagick 6 we need a different regex
       if packages.empty?
-        packages = `pkg-config --list-all`.scan(/(ImageMagick) .*/).flatten
+        # ImageMagick 6.7 does not have package file like ImageMagick-6.Q16.pc
+        packages = detect_imagemagick_packages(['ImageMagick'])
       end
 
       if packages.empty?
         exit_failure "Can't install RMagick #{RMAGICK_VERS}. Can't find ImageMagick with pkg-config\n"
       end
 
-      if packages.length > 1
-
-        im7_packages = packages.grep(/\AImageMagick-7/)
-
-        if im7_packages.any?
-          checking_for('forced use of ImageMagick 6') do
-            if ENV['USE_IMAGEMAGICK_6']
-              packages -= im7_packages
-              true
-            else
-              packages = im7_packages
-              false
-            end
+      if im6_package_exists? && im7_package_exists?
+        checking_for('forced use of ImageMagick 6') do
+          if ENV['USE_IMAGEMAGICK_6']
+            packages = detect_im6_packages
```

Comment:
Not sure if this is an issue but there seems to be something special for `ImageMagick 6.7`? Are you now not overwriting that here when `USE_IMAGEMAGICK_6` is set?

Reply:
If only ImageMagick 6.7 is installed, it will not pass through here.
ImageMagick 6.7 is quite old, so I think it would be best to avoid using it as much as possible.